stakedb/ticketpool.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2018, The dcrdata developers.
+// See LICENSE for details.
+
+package stakedb
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/asdine/storm"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+// TicketPool contains the live ticket pool diffs (tickets in/out) between
+// adjacent block heights in a chain. Diffs are applied in sequence by inserting
+// and removing ticket hashes from a pool, represented as a map. A []PoolDiff
+// stores these diffs, with a cursor pointing to the next unapplied diff. An
+// on-disk database of diffs is maintained using the storm wrapper for boltdb.
+type TicketPool struct {
+	*sync.RWMutex
+	cursor int64
+	tip    int64
+	diffs  []PoolDiff
+	pool   map[chainhash.Hash]struct{}
+	diffDB *storm.DB
+}
+
+// PoolDiff represents the tickets going in and out of the live ticket pool from
+// one height to the next.
+type PoolDiff struct {
+	In  []chainhash.Hash
+	Out []chainhash.Hash
+}
+
+// PoolDiffDBItem is the type in the storm live ticket DB. The primary key (id)
+// is Height.
+type PoolDiffDBItem struct {
+	Height   int64 `storm:"id"`
+	PoolDiff `storm:"inline"`
+}
+
+// NewTicketPool constructs a TicketPool by opening the persistent diff db,
+// loading all known diffs, initializing the TicketPool values.
+func NewTicketPool(dbFile string) (*TicketPool, error) {
+	// Open ticket pool diffs database
+	db, err := storm.Open(dbFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed storm.Open: %v", err)
+	}
+
+	// Load all diffs
+	var poolDiffs []PoolDiffDBItem
+	err = db.AllByIndex("Height", &poolDiffs)
+	if err != nil {
+		return nil, fmt.Errorf("failed (*storm.DB).AllByIndex: %v", err)
+	}
+	diffs := make([]PoolDiff, len(poolDiffs))
+	for i := range poolDiffs {
+		diffs[i] = poolDiffs[i].PoolDiff
+	}
+
+	// Construct TicketPool with loaded diffs and diff DB
+	return &TicketPool{
+		RWMutex: new(sync.RWMutex),
```

Comment:
meh, not needed.

Reply:
What's not needed?  Thread-safing this is needed, so do you mean the `new`?  It's a pointer to the mutex so it has to be allocated manually.